File: src/navigation/route_planner.py
import json
from pathlib import Path
from dataclasses import dataclass, field, asdict
from typing import List, Optional, Tuple
from datetime import datetime

from src.mavlink.telemetry import LatLon
from src.navigation.calculations import (
    haversine_distance, bearing_to, cross_track_distance,
    along_track_distance, eta_seconds
)


@dataclass
class Waypoint:
    id: int
    lat: float
    lon: float
    altitude: float
    radius: float = 50.0
    action: str = "FLYTHROUGH"
    orbit_radius: float = 100.0
    orbit_turns: int = 1
    target_altitude: float = 0.0

    def to_latlon(self) -> LatLon:
        return LatLon(self.lat, self.lon)

    def get_action_name(self) -> str:
        names = {
            "FLYTHROUGH": "Пролёт",
            "ORBIT_ALTITUDE": "Кружить до высоты",
            "ORBIT_TURNS": "Кружить N кругов",
            "ORBIT_INFINITE": "Кружить бесконечно",
        }
        return names.get(self.action, self.action)


@dataclass
class Route:
    name: str
    waypoints: List[Waypoint] = field(default_factory=list)
    created: str = ""
    modified: str = ""

    def __post_init__(self):
        if not self.created:
            self.created = datetime.now().isoformat()
        if not self.modified:
            self.modified = self.created


class RoutePlanner:
    def __init__(self):
        self._route: Optional[Route] = None
        self._active_waypoint_idx = 0

    def load_route(self, path: Path) -> Optional[Route]:
        try:
            with open(path, 'r', encoding='utf-8') as f:
                data = json.load(f)

            waypoints = []
            for wp_data in data.get('waypoints', []):
                wp = Waypoint(
                    id=wp_data.get('id', len(waypoints) + 1),
                    lat=wp_data['lat'],
                    lon=wp_data['lon'],
                    altitude=wp_data.get('altitude', wp_data.get('alt', 100)),
                    radius=wp_data.get('radius', 50.0),
                    action=wp_data.get('action', 'FLYTHROUGH'),
                    orbit_radius=wp_data.get('orbit_radius', 100.0),
                    orbit_turns=wp_data.get('orbit_turns', 1),
                    target_altitude=wp_data.get('target_altitude', 0.0)
                )
                waypoints.append(wp)

            self._route = Route(
                name=data.get('name', path.stem),
                waypoints=waypoints,
                created=data.get('created', ''),
                modified=data.get('modified', '')
            )
            self._active_waypoint_idx = 0
            return self._route

        except Exception as e:
            print(f"Failed to load route: {e}")
            return None

    def save_route(self, route: Route, path: Path) -> bool:
        try:
            route.modified = datetime.now().isoformat()

            data = {
                'name': route.name,
                'created': route.created,
                'modified': route.modified,
                'waypoints': [
                    {
                        'id': wp.id,
                        'lat': wp.lat,
                        'lon': wp.lon,
                        'altitude': wp.altitude,
                        'radius': wp.radius,
                        'action': wp.action,
                        'orbit_radius': wp.orbit_radius,
                        'orbit_turns': wp.orbit_turns,
                        'target_altitude': wp.target_altitude
                    }
                    for wp in route.waypoints
                ]
            }

            path.parent.mkdir(parents=True, exist_ok=True)
            with open(path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2)

            return True
        except Exception as e:
            print(f"Failed to save route: {e}")
            return False

    def get_route(self) -> Optional[Route]:
        return self._route

    def clear_route(self):
        self._route = None
        self._active_waypoint_idx = 0

    def clear_waypoints(self):
        if self._route:
            self._route.waypoints = []
            self._active_waypoint_idx = 0

    def create_route(self, name: str) -> Route:
        self._route = Route(name=name, waypoints=[])
        self._active_waypoint_idx = 0
        return self._route

    def add_waypoint(self, lat: float, lon: float, altitude: float = 100.0,
                     radius: float = 50.0, action: str = "FLYTHROUGH",
                     orbit_radius: float = 100.0, orbit_turns: int = 1,
                     target_altitude: float = 0.0) -> Optional[Waypoint]:
        if not self._route:
            return None
        new_id = len(self._route.waypoints) + 1
        wp = Waypoint(
            id=new_id,
            lat=lat,
            lon=lon,
            altitude=altitude,
            radius=radius,
            action=action,
            orbit_radius=orbit_radius,
            orbit_turns=orbit_turns,
            target_altitude=target_altitude
        )
        self._route.waypoints.append(wp)
        return wp

    def remove_waypoint(self, index: int) -> bool:
        if not self._route or index < 0 or index >= len(self._route.waypoints):
            return False
        self._route.waypoints.pop(index)
        for i, wp in enumerate(self._route.waypoints):
            wp.id = i + 1
        if self._active_waypoint_idx >= len(self._route.waypoints):
            self._active_waypoint_idx = max(0, len(self._route.waypoints) - 1)
        return True

    def get_active_waypoint(self) -> Optional[Waypoint]:
        if not self._route or not self._route.waypoints:
            return None
        if self._active_waypoint_idx >= len(self._route.waypoints):
            return None
        return self._route.waypoints[self._active_waypoint_idx]

    def get_active_waypoint_index(self) -> int:
        return self._active_waypoint_idx

    def get_waypoint_count(self) -> int:
        if not self._route:
            return 0
        return len(self._route.waypoints)

    def next_waypoint(self) -> Optional[Waypoint]:
        if not self._route or not self._route.waypoints:
            return None
        if self._active_waypoint_idx < len(self._route.waypoints) - 1:
            self._active_waypoint_idx += 1
        return self.get_active_waypoint()

    def prev_waypoint(self) -> Optional[Waypoint]:
        if not self._route or not self._route.waypoints:
            return None
        if self._active_waypoint_idx > 0:
            self._active_waypoint_idx -= 1
        return self.get_active_waypoint()

    def set_active_waypoint(self, index: int):
        if self._route and 0 <= index < len(self._route.waypoints):
            self._active_waypoint_idx = index

    def is_waypoint_reached(self, position: LatLon) -> bool:
        wp = self.get_active_waypoint()
        if not wp:
            return False
        distance = haversine_distance(position.lat, position.lon, wp.lat, wp.lon)
        return distance <= wp.radius

    def distance_to_waypoint(self, position: LatLon) -> float:
        wp = self.get_active_waypoint()
        if not wp:
            return float('inf')
        return haversine_distance(position.lat, position.lon, wp.lat, wp.lon)

    def bearing_to_waypoint(self, position: LatLon) -> float:
        wp = self.get_active_waypoint()
        if not wp:
            return 0.0
        return bearing_to(position.lat, position.lon, wp.lat, wp.lon)

    def eta_to_waypoint(self, position: LatLon, groundspeed: float) -> float:
        distance = self.distance_to_waypoint(position)
        return eta_seconds(distance, groundspeed)

    def cross_track_error(self, position: LatLon) -> float:
        if not self._route or len(self._route.waypoints) < 2:
            return 0.0
        if self._active_waypoint_idx == 0:
            return 0.0

        wp1 = self._route.waypoints[self._active_waypoint_idx - 1]
        wp2 = self._route.waypoints[self._active_waypoint_idx]

        return cross_track_distance(
            position.lat, position.lon,
            wp1.lat, wp1.lon,
            wp2.lat, wp2.lon
        )

    def get_previous_waypoint(self) -> Optional[Waypoint]:
        if not self._route or self._active_waypoint_idx == 0:
            return None
        return self._route.waypoints[self._active_waypoint_idx - 1]

    def get_waypoints_for_display(self) -> List[dict]:
        if not self._route:
            return []
        return [
            {
                'lat': wp.lat,
                'lon': wp.lon,
                'id': wp.id,
                'altitude': wp.altitude,
                'action': wp.action,
                'action_name': wp.get_action_name(),
                'radius': wp.radius,
                'orbit_radius': wp.orbit_radius,
                'orbit_turns': wp.orbit_turns,
                'target_altitude': wp.target_altitude
            }
            for wp in self._route.waypoints
        ]

    def is_route_complete(self) -> bool:
        if not self._route:
            return True
        return self._active_waypoint_idx >= len(self._route.waypoints)

File: src/navigation/RoutePlanner.cpp
#include "RoutePlanner.h"
#include <spdlog/spdlog.h>
#include <fstream>
#include <filesystem>
#include <chrono>

namespace vtol {

namespace {
std::string nowIso() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    std::tm tm{};
#ifdef _WIN32
    localtime_s(&tm, &time);
#else
    localtime_r(&time, &tm);
#endif
    char buf[64];
    std::strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", &tm);
    return buf;
}
} // anonymous

std::optional<Route> RoutePlanner::loadRoute(const std::string& path)
{
    try {
        std::ifstream file(path);
        if (!file.is_open()) {
            SPDLOG_ERROR("Failed to open route file: {}", path);
            return std::nullopt;
        }

        nlohmann::json data;
        file >> data;

        Route route;
        route.name = data.value("name", std::filesystem::path(path).stem().string());
        route.created = data.value("created", "");
        route.modified = data.value("modified", "");

        for (const auto& wpData : data.value("waypoints", nlohmann::json::array())) {
            Waypoint wp;
            wp.id = wpData.value("id", static_cast<int>(route.waypoints.size()) + 1);
            wp.lat = wpData.at("lat").get<double>();
            wp.lon = wpData.at("lon").get<double>();
            wp.altitude = wpData.contains("altitude") ? wpData["altitude"].get<double>()
                        : wpData.value("alt", 100.0);
            wp.radius = wpData.value("radius", 150.0);
            wp.action = wpData.value("action", "FLYTHROUGH");
            wp.orbit_radius = wpData.value("orbit_radius", 150.0);
            wp.orbit_turns = wpData.value("orbit_turns", 1);
            wp.climb_enroute = wpData.value("climb_enroute", false);
            route.waypoints.push_back(wp);
        }

        m_route = route;
        m_activeIdx = 0;
        SPDLOG_INFO("Route loaded: {} ({} waypoints)", route.name, route.waypoints.size());
        return route;

    } catch (const std::exception& e) {
        SPDLOG_ERROR("Failed to load route: {}", e.what());
        return std::nullopt;
    }
}

bool RoutePlanner::saveRoute(const Route& route, const std::string& path)
{
    try {
        namespace fs = std::filesystem;
        fs::create_directories(fs::path(path).parent_path());

        nlohmann::json data;
        data["name"] = route.name;
        data["created"] = route.created;
        data["modified"] = nowIso();

        auto& wps = data["waypoints"] = nlohmann::json::array();
        for (const auto& wp : route.waypoints) {
            wps.push_back({
                {"id", wp.id}, {"lat", wp.lat}, {"lon", wp.lon},
                {"altitude", wp.altitude}, {"radius", wp.radius},
                {"action", wp.action}, {"orbit_radius", wp.orbit_radius},
                {"orbit_turns", wp.orbit_turns}, {"climb_enroute", wp.climb_enroute}
            });
        }

        std::ofstream file(path);
        if (!file.is_open()) {
            SPDLOG_ERROR("Failed to write route to {}", path);
            return false;
        }
        file << data.dump(2);
        SPDLOG_INFO("Route saved to {}", path);
        return true;

    } catch (const std::exception& e) {
        SPDLOG_ERROR("Failed to save route: {}", e.what());
        return false;
    }
}

Route& RoutePlanner::createRoute(const std::string& name)
{
    m_route = Route{name, {}, nowIso(), ""};
    m_activeIdx = 0;
    return *m_route;
}

Waypoint* RoutePlanner::addWaypoint(double lat, double lon, double alt,
                                      double radius, const std::string& action,
                                      double orbitRadius, int orbitTurns, bool climbEnroute)
{
    if (!m_route) return nullptr;
    int newId = static_cast<int>(m_route->waypoints.size()) + 1;
    m_route->waypoints.push_back({newId, lat, lon, alt, radius, action, orbitRadius, orbitTurns, climbEnroute});
    return &m_route->waypoints.back();
}

bool RoutePlanner::removeWaypoint(int index)
{
    if (!m_route || index < 0 || index >= static_cast<int>(m_route->waypoints.size()))
        return false;
    m_route->waypoints.erase(m_route->waypoints.begin() + index);
    // Reindex
    for (int i = 0; i < static_cast<int>(m_route->waypoints.size()); ++i)
        m_route->waypoints[i].id = i + 1;
    if (m_activeIdx >= static_cast<int>(m_route->waypoints.size()))
        m_activeIdx = std::max(0, static_cast<int>(m_route->waypoints.size()) - 1);
    return true;
}

bool RoutePlanner::updateWaypoint(int index, const Waypoint& newData)
{
    if (!m_route || index < 0 || index >= static_cast<int>(m_route->waypoints.size()))
        return false;
    int savedId = m_route->waypoints[index].id;
    m_route->waypoints[index] = newData;
    m_route->waypoints[index].id = savedId;  // keep sequential id
    return true;
}

bool RoutePlanner::moveWaypoint(int fromIndex, int toIndex)
{
    if (!m_route) return false;
    int n = static_cast<int>(m_route->waypoints.size());
    if (fromIndex < 0 || fromIndex >= n || toIndex < 0 || toIndex >= n) return false;
    if (fromIndex == toIndex) return true;

    Waypoint wp = m_route->waypoints[fromIndex];
    m_route->waypoints.erase(m_route->waypoints.begin() + fromIndex);
    m_route->waypoints.insert(m_route->waypoints.begin() + toIndex, wp);

    // Reindex ids
    for (int i = 0; i < static_cast<int>(m_route->waypoints.size()); ++i)
        m_route->waypoints[i].id = i + 1;

    // Adjust active index to follow the moved waypoint if it was active
    if (m_activeIdx == fromIndex)
        m_activeIdx = toIndex;
    else if (fromIndex < toIndex && m_activeIdx > fromIndex && m_activeIdx <= toIndex)
        --m_activeIdx;
    else if (fromIndex > toIndex && m_activeIdx >= toIndex && m_activeIdx < fromIndex)
        ++m_activeIdx;

    return true;
}

Waypoint* RoutePlanner::activeWaypoint()
{
    if (!m_route || m_route->waypoints.empty()) return nullptr;
    if (m_activeIdx >= static_cast<int>(m_route->waypoints.size())) return nullptr;
    return &m_route->waypoints[m_activeIdx];
}

const Waypoint* RoutePlanner::activeWaypoint() const
{
    if (!m_route || m_route->waypoints.empty()) return nullptr;
    if (m_activeIdx >= static_cast<int>(m_route->waypoints.size())) return nullptr;
    return &m_route->waypoints[m_activeIdx];
}

Waypoint* RoutePlanner::nextWaypoint()
{
    if (!m_route || m_route->waypoints.empty()) return nullptr;
    if (m_activeIdx < static_cast<int>(m_route->waypoints.size()) - 1)
        ++m_activeIdx;
    return activeWaypoint();
}

Waypoint* RoutePlanner::prevWaypoint()
{
    if (!m_route || m_route->waypoints.empty()) return nullptr;
    if (m_activeIdx > 0) --m_activeIdx;
    return activeWaypoint();
}

void RoutePlanner::setActiveWaypoint(int index)
{
    if (m_route && index >= 0 && index < static_cast<int>(m_route->waypoints.size()))
        m_activeIdx = index;
}

bool RoutePlanner::isWaypointReached(const LatLon& pos) const
{
    auto* wp = activeWaypoint();
    if (!wp) return false;
    return nav::haversineDistance(pos.lat, pos.lon, wp->lat, wp->lon) <= wp->radius;
}

double RoutePlanner::distanceToWaypoint(const LatLon& pos) const
{
    auto* wp = activeWaypoint();
    if (!wp) return std::numeric_limits<double>::infinity();
    return nav::haversineDistance(pos.lat, pos.lon, wp->lat, wp->lon);
}

double RoutePlanner::bearingToWaypoint(const LatLon& pos) const
{
    auto* wp = activeWaypoint();
    if (!wp) return 0.0;
    return nav::bearingTo(pos.lat, pos.lon, wp->lat, wp->lon);
}

double RoutePlanner::etaToWaypoint(const LatLon& pos, double groundspeed) const
{
    return nav::etaSeconds(distanceToWaypoint(pos), groundspeed);
}

double RoutePlanner::crossTrackError(const LatLon& pos) const
{
    if (!m_route || m_route->waypoints.size() < 2 || m_activeIdx == 0)
        return 0.0;
    const auto& wp1 = m_route->waypoints[m_activeIdx - 1];
    const auto& wp2 = m_route->waypoints[m_activeIdx];
    return nav::crossTrackDistance(pos.lat, pos.lon, wp1.lat, wp1.lon, wp2.lat, wp2.lon);
}

const Waypoint* RoutePlanner::previousWaypoint() const
{
    if (!m_route || m_activeIdx == 0) return nullptr;
    return &m_route->waypoints[m_activeIdx - 1];
}

bool RoutePlanner::isRouteComplete() const
{
    if (!m_route) return true;
    return m_activeIdx >= static_cast<int>(m_route->waypoints.size());
}

} // namespace vtol
